Elastic Client Adapter: initial implementation

joxl opened the pull request:
## Technical Summary

New Elastic "client adapter" layer which will eventually replace `corehq.elastic` and `corehq.util.es.interface` in order to support Elastic index "multiplexing".  Documentation for this feature is outlined in the [Adapter Design section](https://github.com/dimagi/commcare-hq/blob/5347073ea0bc504458928c4f282b2dbe3067e038/corehq/apps/es/README.rst#adapter-design) of the HQ app README.

This PR is the first step towards that goal. It accomplishes a few initial objectives:

- define the client adapter role and usage patterns
- unit tests for all new client adapter logic
- complete reST documentation in high-level `.rst` files as well comprehensive docstrings
- "rewire" aforementioned modules (`corehq.elastic` and `corehq.util.es.interface`) to proxy through the new implementation in order to facilitate incremental code changes

Ticket: [SAAS-13171](https://dimagi-dev.atlassian.net/browse/SAAS-13171)

#### Code Review :tropical_fish:

Commit-by-commit is recommended due to several narrow-scope commits with wide-reaching effects (keyword argument renames, etc). The `corehq.apps.es.client` (and its test) modules, which are new, may be easiest to review unified:

- unified [corehq/apps/es/client.py](https://github.com/dimagi/commcare-hq/pull/31125/files#diff-d0e563f9edf0e4a5314a12385010f4b66cf8e4ba3ead9d9a5ea831ceabb9476f)
- unified [corehq/apps/es/tests/test_client.py](https://github.com/dimagi/commcare-hq/pull/31125/files#diff-f6214f325abf0a594c94bb9a2448d12004ac0f5b6573790330e8c6fcc39a7067)

**Note**: the `TODO` comments in this PR will be resolved during future work on the adapter.

#### Documentation

Our readthedocs.io account only builds docs for `master`. If you wish to view the rendered docs as composed in this PR, you can build and view them locally:

```bash
git fetch origin
git checkout -b trashme --no-track origin/jm/elastic-client-adapter
cd ./docs
make html
open ./_build/html/elasticsearch.html
```

## Safety Assurance

### Safety story

Many new tests, all Elasticsearch tests are passing. Some existing tests required updating.  This is a high-risk change because the blast radius is large.  There are large portions of pillowtop and reporting code using both the existing "interface" modules, as well as discrete logic with the low-level Elasticsearch 3rd party client library directly.  The changes in this PR are supposed to be transparent to those areas.

### Automated test coverage

Over 100 new Elasticsearch tests. The delta is slightly lower because several obsolete tests have been removed.

### QA Plan

QA ticket: [QA-3949](https://dimagi-dev.atlassian.net/browse/QA-3949)

QA plan will include everything in their "Elasticsearch arsenal", e.g.

- submissions and report verification on all Elasticsearch entities (domains, apps, forms, cases, users, groups, etc)
- user create/update changes
- pillows performance benchmarking
- probably more

### Migrations

No migrations

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

joxl committed:
Update imports moved to new `corehq.apps.es.const` module

joxl committed:
Replace `es_instance_alias` kwarg with boolean `for_export`

joxl committed:
Update Elasticsearch document structure and terminology

joxl committed:
Initial client adapter implementation

joxl committed:
Use constants and rename `ES_SETTINGS` to `ELASTIC_SETTINGS_OVERRIDE`

- replace static strings with constants in `test_prod_es_indices.py`
- `ES_SETTINGS` (and tests it is used for) are very brittle and will
  break reindexing efforts. This change makes note of that and renames
  things to be more "visible".

joxl committed:
Define skeleton document adapters

- new document adapter for each Elastic index
- index names are defined in `settings.py`
- new settingshelper for updating index names during tests

joxl committed:
Update pillow mapping constants to source from document adapters

Values provided by document adapters:

- index
- doc_type

joxl committed:
Update `corehq.elastic` to use Elastic client adapters

joxl committed:
Update prune_elastic_indices manage command to use ElasticManageAdapter

joxl committed:
Begin "rerouting" interface.py through Elastic client adapters

joxl committed:
Change `get_aliases()` method to return dict of lists instead of `defaultdict(set)`

joxl committed:
Adapter conversion: ElasticsearchInterface.get_aliases()

joxl committed:
Adapter conversion: ElasticsearchInterface.put_mapping()

joxl committed:
Add `index_configure_for_...()` methods to ElasticManageAdapter

joxl committed:
Adapter conversion: ElasticsearchInterface.update_index_settings_...()

joxl committed:
Adapter conversion: ElasticsearchInterface.doc_exists()

joxl committed:
Adapter conversion: ElasticsearchInterface.get_doc()

joxl committed:
Add transient `transform()` methods on new document adapters

joxl committed:
Add mapping to ElasticDocumentAdapter "base requirements"

Earlier than planned, transient function implementation until mappings
get moved. Needed to write "modern" tests.

joxl committed:
Adapter conversion: ElasticsearchInterface.get_bulk_docs()

joxl committed:
Write test to verify ambiguous use of Elastic Get API

Add ElasticDocumentAdapter docs to clarify the importance of the Elastic
`_type` meta property.

joxl committed:
Fix ESView _type bug

joxl committed:
Adapter conversion: ElasticsearchInterface.index_doc()

- remove failing tests that are now redundant (covered by adapter tests)
- fix fragile tests that index documents without meeting adapter constraints

joxl committed:
Adapter conversion: ElasticsearchInterface.update_doc_fields()

joxl committed:
Adapter conversion: ElasticsearchInterface.count()

joxl committed:
Adapter conversion: ElasticsearchInterface.delete_doc()

joxl committed:
Adapter conversion: ElasticsearchInterface.search()

joxl committed:
Adapter conversion: ElasticsearchInterface.iter_scroll()

joxl committed:
Update delete_case_search_cases() to use ElasticCaseSearch.bulk_delete()

ElasticsearchInterface.bulk_ops() is going away

joxl committed:
Add document source verification and improve tests

joxl committed:
Overhaul bulk processing logic

joxl committed:
Adapter conversion: ElasticsearchInterface.bulk_ops()

joxl committed:
Begin groundwork for new elastic client layer

joxl committed:
Naming correction `s/results/result/g`

joxl committed:
Remove unused `query` property

millerdev commented on corehq/apps/es/client.py:
```diff
@@ -0,0 +1,933 @@
+class ElasticDocumentAdapter(ElasticClientAdapter):
+    """Base for subclassing document-specific adapters."""
+
+    index_key = None  # set by subclass
+    type = None  # set by subclass
+    mapping = None  # set by subclass
```
Rather than setting these to `None` (which is not a valid value), what do you think of adding a note in the docstring that these must be set, and not actually set them here on this base class.

millerdev reviewed (commented):
Reviewed collaboratively with @joxl commits up to https://github.com/dimagi/commcare-hq/pull/31125/commits/ba1e6b021c6db84ef5a7da39fcc9377a0d599178 (that commit was only partially reviewed).

Review to be continued tomorrow.

joxl committed:
Improve shard failure checking and tests

joxl committed:
Misc documentation cleanup and improvements

calellowitz commented:
A few high level questions:

Is there a design doc for this work that lays out the changes made and rationale for those decisions? Having a bit more context would make it much easier to review.

You mention "Elastic index multiplexing" as one of the key features this unlocks, what do you mean by that phrase and how does this work allow it?

Is it possible to break this into more bite sized PRs, much like Daniel is doing with the dbaccessor work? A number of people will probably want to review this work since it is so large and cross cutting, and that may make it easier for everyone, especially those who were not involved in earlier conversations.

millerdev reviewed (commented):
Full pass completed.

joxl committed:
Code review: fix docstring terminations

joxl committed:
Code review: improve `get_client()` tests

joxl committed:
Code review: fix default keyword arg regression in esaccessors.py

joxl committed:
Code review: future-proof manage adapter explanation

joxl committed:
Code review: improve `cluster_routing()` semantics and fix tests

joxl committed:
Code review: enforce `indices_refresh()` arg type

joxl committed:
Code review: move `index_refresh()` up and add missing test

joxl committed:
Code review: use docstring to document requirements rather than invalid values

joxl committed:
Code review: improve ping fail test

joxl committed:
Code review: add docstring notes for legacy interface grepability

joxl committed:
Code review: improve `search()` docstring formatting

joxl committed:
Code review: add tests to ensure bulk methods support generators

joxl commented on corehq/apps/es/client.py:
```diff
@@ -0,0 +1,935 @@
+        :raises: ``ESShardFailure``, ``ValueError``
+        """
+        if not isinstance(result, dict):
+            raise ValueError(f"invalid Elastic result object: {result}")
```
:ballot_box_with_check:

joxl committed:
Code review: use Enum for BulkActionItem op_type

joxl commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -0,0 +1,1150 @@
+        def get_search_hits():
+            return doc_adapter.search({})["hits"]["hits"]
+
+        self.assertEqual([], get_search_hits())
```
Wow, good catch. Yes, it can. It seems extremely unlikely, but is in fact possible. I'll fix this.

joxl committed:
Code review: fix `test_indices_refresh` test to ensure it never flakes

joxl committed:
Code review: perform real flush as part of `test_index_flush`

joxl committed:
Code review: don't memoize document adapter class properties

joxl committed:
Code review: fix previously missed `params` redundancy issue

joxl committed:
Code review: rename transform* functions and methods

- `transform()` --> `from_python()`
- `transform_full()` --> `to_json()`
- `transform_dict_with_possible_id()` --> `from_dict_with_possible_id()`

joxl committed:
Improve adapter (and testcase) class names

joxl committed:
Remove unused `reindex()` method

joxl committed:
Add test for condition where `ELASTICSEARCH_HOST` is ignored

joxl committed:
Code review: move Elastic index names out of `settings.py`

joxl committed:
Fixup: `get_last_submission_time_for_users()`

- set `for_export=True` to match behavior prior to c1ba2a51b54

joxl committed:
Code review: improve Elastic host parser and add tests

millerdev commented on corehq/apps/es/client.py:
```diff
@@ -0,0 +1,994 @@
+        return self.op_type is self.OpType.index
+
+    def __repr__(self):
+        human_op = getattr(self.op_type, "name", "???")  # avoid AttributeError
```
nit: In what situation does `self.op_type.name` not exist?
```suggestion
        human_op = self.op_type.name
```

millerdev commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -346,14 +346,27 @@ def test_indices_refresh(self):
+            self.adapter.indices_refresh([doc_adapter.index])
+            docs = [h["_source"] for h in get_search_hits()]
+        finally:
+            set_refresh_interval("5s")  # set it back to previous value
```
nit: How does this test know what the previous value was? And what if it changes?

If this is a pattern that may be useful in other tests it might make sense to extract it to a context manager utility function, which would be used like
```py
with suspend_view_refresh(self.adapter):
    doc_adapter.upsert(TestDoc("1", "test"))
    self.assertEqual([], get_search_hits())
    ...
```

joxl replied:
It doesn't.  In fact the test doesn't need to set anything back because the index is destroyed immediately after (this is an index-specific setting).  I'm going to remove this.

snopoke reviewed (commented):
@joxl is there a spec or other doc that gives more background to this PR. I'm struggling to get my head around what we're trying to accomplish with this change and why we need a 3rd level of abstraction on top of ES (ES library, `ElasticsearchInterface`, and now this).

joxl commented:
Hey @snopoke, I'm putting together a high-level doc. This will not be another abstraction layer, it is replacing `ElasticsearchInterface` (mentioned in the summary above).

joxl committed:
Rename `ElasticDocumentAdapter.index` to `.index_name`

joxl committed:
Code review: improve tests

- `test_index_exists`: remove unnecessary cleanup boilerplate (index is
  cleaned up by `tearDown()` method)
- `test_indices_refresh`: remove unnecessary try/finally block (index is
  deleted after test)
- `test_index_put_alias_flips_existing`: replace boilerplate with new
  contextmanager

joxl committed:
Rename `get_mapping` to `get_adapter_mapping`

- decreases boilerplate and verbosity in next commit

joxl committed:
Code review: rename `ElasticDocumentAdapter.upsert()` to `.index()`

- new `index_name` class attribute releases `index` name
- resolves inconsistency between BulkActionItem and ElasticDocumentAdapter

joxl committed:
Code review: improve `_render_bulk_action` test

- eliminates unnecessary `getattr(...)` in BulkActionItem repr method

joxl committed:
Revert terminology update

- "indexes" is not grammatically incorrect and may be preferred for code
  searchability
- reverts subset of changes from: dba33f5e2d69546f97c105e24301d8a77a99bb85

joxl committed:
Add Elastic adapter design documentation

joxl committed:
Move reST documentation out of `corehq.apps.es.client` module docstring

- reST structural elements don't work well in Python docstrings
- see: https://stackoverflow.com/questions/62855567/headings-when-using-autodoc-in-sphinx#62856198

joxl commented:
@calellowitz  and @snopoke, design doc now available [here](https://github.com/dimagi/commcare-hq/blob/jm/elastic-client-adapter/corehq/apps/es/README.rst).

joxl committed:
Doc formatting fix (reST != md)

joxl committed:
Fix doc code example (`datetime.date`) and improve formatting

joxl committed:
Documentation final review pass improvements

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+other:
+
+1. Merge and deploy a PR that configures multiplexing on an index.
+2. Execute an idempotent management command that updates the secondary index
```
Does such a command exist?

joxl replied:
No.

joxl replied:
> This seems to be the crux of the update process

You're absolutely right.

> Do you have a sense how this command will work?

I have the design worked out but not implemented. This will come after the multiplexing implementation because it depends on how that logic handles differences between the primary and secondary indexes (e.g. multiplexer gets a DELETE request, doc exists in primary but not secondary yet, etc). This command will initiate the actual /reindex API request, monitor that task's progress and verify both indexes are equal before returning and reporting the success. Kind of like rsync, but for an Elastic index, and with fewer options.

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
```
One of the reasons to have the back-filling run as a separate process was to prevent any potential lag / performance issues with maintaining the current index. Doing the multiplexing certainly makes the process simpler but do we have any options if the additional operations do cause lag?

joxl replied:
We need measurements to know how multiplexing will impact the cluster. There is an ongoing discussion with QA on benchmarking HQ Elasticsearch performance with the specific goal of being able to measure any performance degradation as a result of multiplexing. See [QA-3843](https://dimagi-dev.atlassian.net/browse/QA-3843)

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+
+    adapter = ElasticManageAdapter()
+    adapter.index_create("books")
+    mapping = {"properties": {
+        "author": {"type": "text"},
+        "title": {"type": "text"},
+        "published": {"type": "date"},
+    }}
```
How does this work if the old and new indexes have different mappings? Or does the management adaptor not support multiplexing? I imagine it would make sense for the management adapter to be tied to a single index otherwise it would make it hard to management them independently.

joxl replied:
Index creation and manipulation for multiplexing is not yet implemented, but it will certainly be done in a "multiplexing-aware" way at that time.

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+  naming between production code and tests.
+- A ``type`` attribute whose value is the name is the Elastic ``_type`` for
+  documents used by the adapter.
+- A ``mapping`` which defines the structure and properties for documents managed
+  by the adapter.
```
What's the purpose of the mapping in the document adapter?

snopoke replied:
Sure, but the document adapters don't need to know about the mapping unless you plan to add query validation or something like that right?

joxl replied:
Not in this PR. The multiplexing implementation will need it on the adapter to support things like mapping changes.

snopoke replied:
I'd like more info on this piece too please :)

snopoke reviewed (commented):
@joxl thanks for the docs. I don't see anything in them about how the multiplexing would work or how the adapters would be configured for multiple indexes.

joxl commented:
> I don't see anything in them about how the multiplexing would work or how the adapters would be configured for multiple indexes.

This PR doesn't implement multiplexing.  It's a "step towards that goal" (see the PR summary).

joxl committed:
Revert Elastic settings override changes

- `s/ELASTIC_SETTINGS_OVERRIDE/ES_SETTINGS/g`
- previously changed in 9c853543840

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

joxl committed:
Update `hqdomains` index name as exists on master branch

joxl committed:
Decorate `ProdIndexManagementTest` as an Elastic test

snopoke commented on corehq/pillows/mappings/user_mapping.py:
```diff
@@ -232,12 +235,10 @@
 
-USER_ES_ALIAS = prefix_for_tests('hqusers')
-
 USER_INDEX_INFO = ElasticsearchIndexInfo(
```
What's the end goal for these? Do you plan to remove them completely? I'm just finding it the crossover between the adapters and these classes a bit weird. I almost think that the adapters should either replace these entirely or just reference them.

joxl replied:
Yes, but this PR doesn't include the replacement, so they need to persist until that's implemented.

snopoke commented on corehq/util/es/interface.py:
```diff
@@ -1,195 +1,99 @@
-        self.es.index(index_alias, doc_type, body=self._without_id_field(doc), id=doc_id,
-                      params=params or {})
+        doc_adapter = self._get_doc_adapter(index_alias, doc_type)
+        if doc.get("_id", object()) != doc_id:
+            doc["_id"] = doc_id
```
This is a confusing check. If `_id` is in the doc and not equal to `doc_id` then I think it would be better to raise and exception instead of just overwriting the value in the doc.

joxl replied:
This is functionally identical to what the previous implementation did. Removed logic was:
```python
self.es.index(index_alias, doc_type, body=self._without_id_field(doc), id=doc_id,
                      params=params or {})
```
Where
- `body=self._without_id_field(doc)` == _"strip `_id` from document"_
- `id=doc_id` == _"explicitly use `doc_id`"_

While fixing the possibility of that being a bug is outside the scope of this commit, you make a good point that this is concerning logic. I'll add a `TODO` here to make sure it gets revisited.

snopoke commented on corehq/util/es/interface.py:
```diff
@@ -1,195 +1,99 @@
-        return results
+        doc_adapter = self._get_doc_adapter(index_alias, doc_type)
+        query = {} if body is None else body
+        params = params if params else {}
```
nit: consistency with the line above and other places in this file
```suggestion
        params = {} if params is not None else params
```

millerdev replied:
Should that have been `params = {} if params is None else params`?

snopoke reviewed (commented):
I notice there is a circular relationship between `corehq.apps.es.registry`, the `ElasticsearchIndexInfo` objects and the `ElasticDocumentAdapter` objects. Is the plan to replace the registry & `ElasticsearchIndexInfo` objects with the adpaters?

joxl commented on corehq/apps/es/case_search.py:
```diff
@@ -130,6 +133,20 @@ def sort_by_case_property(self, case_property_name, desc=False):
+class ElasticCaseSearch(ElasticDocumentAdapter):
+
+    _index_name = "case_search_2018-05-29"
+    type = ElasticCase.type
```
The [previous mapping implementation](https://github.com/dimagi/commcare-hq/blob/77680695c8a0f3134884964410dae14183c45f35/corehq/pillows/mappings/case_search_mapping.py#L16) did not have them separate, so I retained that logic:

**file**: `corehq/pillows/mappings/case_search_mapping.py`

```python
# snip ...
from corehq.pillows.mappings.case_mapping import CASE_ES_TYPE

# snip ...

CASE_SEARCH_INDEX_INFO = ElasticsearchIndexInfo(
    # snip ...
    type=CASE_ES_TYPE,

# snip ...
```

joxl commented:
> Is the plan to replace the registry & ElasticsearchIndexInfo objects with the adpaters?

Yes, that's correct. Adapters will replace both of these.

joxl committed:
Update corehq/util/es/interface.py

Review improvement

Co-authored-by: Simon Kelly <skelly@dimagi.com>

joxl committed:
Code review: add comment about document id in update

millerdev commented on corehq/util/es/interface.py:
```diff
@@ -1,195 +1,103 @@
-        return results
+        doc_adapter = self._get_doc_adapter(index_alias, doc_type)
+        query = {} if body is None else body
+        params = {} if params is not None else params
```
I think this should be
```suggestion
        params = {} if params is None else params
```

joxl replied:
Yes, new one is certainly broken. Upon further inspection, this method doesn't need to do anything with the `params` keyword arg, and it can be removed completely.

joxl committed:
Allow `search(..., params)` keyword arg to pass through via `**kwargs`

- `ElasticDocumentAdapter.search(...)` does not use the `params` arg, so
  allow it to pass directly to the underlying client like any other
  extra keyword arg.

joxl commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -0,0 +1,1255 @@
+from ..exceptions import ESShardFailure, TaskError, TaskMissing
+
+
+@override_settings(ELASTICSEARCH_HOSTS=["localhost"],
```
Yes, good catch.

joxl committed:
Code review: discard memoized clients after settings override tests

joxl committed:
Fix "upsert" references missed in 280d67cf6a8

calellowitz reviewed (commented):
Thanks for the docs and for bearing with me all this time, while I wrapped my head around this work. In general, these changes look fine to me.

I left a few comments and questions in the code, as usual those prefaced with `nit` are less important and mostly stylistic. 

I had a couple higher level questions as well. 

1) Can you explain a bit more how this relates to the multiplexing work? I see that when Simon asked a similar question, you mentioned that was not included in this PR, but it seems to be one of the main rationales for this work. How does this PR enable (or make easier) multiplexing?

2) Do you have a sense how far along this work is? It is hard for me to tell if this represents 10% or 90% or anywhere in between of the changes you plan to make to elasticsearch to facilitate the upgrades. In that vein, are the future plans for this work available anywhere I and others can see them?

3) Do you have a plan to communicate these changes to the broader team? Both those in the current PR and the new ones coming. Many developers use elasticsearch outside of SaaS, and if there are new ways to do operations they may be used to doing the old way, it would be great to proactively let everyone know.

joxl committed:
Code review: improve ElasticDocumentAdapter testing performance

- move ElasticDocumentAdapter tests that don't require an index into a
  separate test class that doesn't perform any index setup
- decreases accumulated test time for the 68 ElasticDocumentAdapter
  tests from 5.96 to 4.82 seconds

joxl committed:
Code review: improve robustness of `temporary_index()` utility

joxl commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -0,0 +1,1255 @@
+        # simply fail all the tests rather than spewing many huge tracebacks
+        self.assertFalse(self.adapter.index_exists(),
+                         f"index exists: {self.adapter.index_name}")
+        self.adapter.create_index({"mappings": {self.adapter.type: self.adapter.mapping}})
```
> Would it not be much faster to only do it once for the whole test class?

In this case it's important that the index get created fresh for each test to enforce that actions from one test cannot affect another.

> a number of the tests don't even make requests to ES

Addressed in 7e3d231ab97 (new commit), the non-index-requiring tests are now split into a separate class that doesn't perform this boilerplate.

joxl commented on corehq/apps/es/tests/utils.py:
```diff
@@ -201,6 +207,23 @@ def decorate(name, decorator):
+        if mapping is None:
+            raise ValueError(f"type_ and mapping args are mutually inclusive "
+                             f"(index={index!r}, type_={type_!r})")
+        manager.index_put_mapping(index, type_, mapping)
```
Nice catch. Addressed in cc8c4d31e85 (with added tests).

millerdev commented on corehq/apps/es/transient_util.py:
```diff
@@ -0,0 +1,103 @@
+    ``from_python()`` logic.
+    """
+    if "_id" in doc:
+        return doc["_id"], {k: doc[k] for k in doc if k != "_id"}
```
Using `del` would mutate the doc, which may be unexpected by calling code. It's a best practice to have functions not mutate arguments passed to them. Or if that is done it should be explicitly documented. I'm not sure if it makes sense to do that in this case? Like, is this always called in a context where the original `doc` dict is thrown away? If yes, then mutating it may be fine?

calellowitz replied:
That is a great point. I hadn't considered if this doc was used elsewhere. My understanding is that it should not be, but if that is not something we can enforce or verify, it may be preferable to keep the existing behavior.

joxl replied:
Yes, avoiding mutation was the goal here.  In the "transient" period of time this function will exist (until the adapters all own their respective `from_python()` logic), this function gets called on documents in many places, similarly to how/where the old `ElasticsearchInterface._without_id_field(doc)` method [did something similar](https://github.com/dimagi/commcare-hq/blob/252e729fda8/corehq/util/es/interface.py#L82-L86).

joxl commented on corehq/apps/es/transient_util.py:
```diff
@@ -0,0 +1,103 @@
+
+    class Adapter(ElasticDocumentAdapter):
+        index_name = index_  # override the classproperty
+        type = type_
```
This is defined within the class scope. It's not overriding a builtin.

joxl commented on corehq/apps/es/transient_util.py:
```diff
@@ -0,0 +1,103 @@
+        mapping_key = (index_info.index, index_info.type)
+        _DOC_MAPPINGS_BY_INDEX[mapping_key] = index_info.mapping
+
+    if is_test:
```
> Can this parameter be handled automatically by settings.UNIT_TESTING?

It could, but it's not for Separation of concerns. The importance of that model in this specific scope is limited, but it makes no difference for other reasons so I defaulted to use that design principle.  This function only gets called from [one place](https://github.com/dimagi/commcare-hq/blob/cc8c4d31e85/corehq/apps/es/app_config.py#L11).

The lookups dicts being built here are for very temporary utility.  I think it's advantageous to keep all the logic in one function for clarity and clean removal later.

calellowitz replied:
Can you explain a bit more how that principle applies here? The function still needs to change behavior based on whether or not it is in a test, and the standard way to do that in HQ is to checking `settings.UNIT_TESTING`. It is highly unusual to pass that as a parameter to a function in hq, and makes it easier to accidentally misuse the function. It seems especially unnecessary to parameterize this since that parameter is hardcoded to the result of the suggested settings value.

I think the mixing of test and production code is generally a code smell, and as mentioned before would prefer the firmer separation, and confining of test only code to the test modules as per our code conventions. That is both because it makes it easier to find the test code, and because it makes it the production code read more cleanly. However, if you feel strongly that the two setups must be in the same function, I would feel a lot better if the separation was enforced via the settings value rather than a supplied parameter so that it is not possible to accidentally run the test setup in production or otherwise misuse this function.

joxl replied:
> Can you explain a bit more how that principle applies here?

Writing functions that change their behavior based on global variables is generally something I try to avoid because it makes code more fragile, often harder to grok, and more challenging to test (e.g. requires extra `mock.patch()` cruft).  This is why I implemented the `settings.UNIT_TESTING` value as an argument rather than making the function reach out into the world and fetch it directly (i.e. based on a general coding practice).  This function is only called in one place, and it would be problematic to call this function from somewhere other than the AppConfig's `ready()` method (I cannot imagine a scenario where doing so would be useful), which is why passing the value as an argument was safe. As of d368f5f4bc4 there is a usage note in this function's docstring as well.

Per request, I've changed this to follow existing conventions in ff7351ded65.

joxl commented on corehq/apps/es/transient_util.py:
```diff
@@ -0,0 +1,103 @@
+                        TEST_ES_MAPPING, TEST_ES_INFO.alias)
+
+
+def _add_test_adapter(descriptor, index_, type_, mapping_, alias):
```
Same as above

>  I think it's advantageous to keep all the logic in one function for clarity and clean removal later.

It fits the [design scope of this module](https://github.com/dimagi/commcare-hq/blob/cc8c4d31e85/corehq/apps/es/transient_util.py#L1-L3) best.

joxl commented on corehq/apps/es/utils.py:
```diff
@@ -1,8 +1,35 @@
 from corehq.util.metrics import metrics_counter
 
 
+class ElasticJSONSerializer(object):
```
Yes.

Disclosure: throughout this PR there has been a (perhaps inconsistent) effort to try avoid using the `ES` acronym since it can cause ambiguities with ECMAScript (this will become more problematic as we move towards Elastic 6 and Elastic 7).

When moving this from the `corehq.elastic` module, I took the opportunity to rename it for that reason.  I'm not embedded on the decision to rename it and I could go both ways.

Further disclosure: I really wish this app was named `corehq.apps.elastic`, but the idea of renaming it broke my invasivity meter.

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+  documents used by the adapter.
+- A ``mapping`` which defines the structure and properties for documents managed
+  by the adapter.
+- A ``from_python()`` classmethod which can convert a Python model object into the
```
Second question first

> is the long term intention to move the full transform from the pillow processor to these adapter classes?

Yes.

> It looks like from_python takes a dictionary rather than a python model class. Is that correct?

By this are you talking about how all the current adapters (e.g. [ElasticUser.from_python()](https://github.com/dimagi/commcare-hq/blob/cc8c4d31e85/corehq/apps/es/users.py#L78-L80)) accept docs as dicts?  Those `from_python()` methods currently accept dicts because at this transient stage of the adapter implementation, pillows still owns and performs the transformation logic from python model instances to dicts. In order to avoid ripping that out of pillows and making this PR even larger, yes, the current _functional_ `from_python()` classmethods accept dicts.  They will not in the future.

calellowitz replied:
All of our current transformation functions from pillows operate on dictionaries rather than classes, does this mean the intention is to change that model?

joxl replied:
Which transformation functions? Functions that transform a document to be indexed _into_ Elasticsearch, or functions that transform a document fetched _from_ Elasticsearch?

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+  JSON-serializable format for writing into the adapter's index.
+
+The combination of ``(index_name, type)`` constrains the document adapter to
+a specific HQ document mapping.  Comparing an Elastic cluster to a Postgres
```
Yes, I'm aware of this, but in order to facilitate a smooth transition from 2.4 (which requires them) to newer a version, they're an unfortunate reality in the interim.

calellowitz replied:
Yeah, sorry I did not mean to suggest you should remove them from this PR. I was wondering if you had a sense how easy it will be to remove them later, and if there are steps we could take, either in enforcing how they are used, or just in documentation that will make it easier to make that change, since that is another key part of the upgrade starting in version 5. That might just be noting in this section for other devs that types are deprecated going forward and new functionality should be developed that rely on them, since this new documentation implies they are a key part of the ES model, which is hopefully only true for the shortest possible time, and we should not count on that for any feature.

joxl replied:
> I was wondering if you had a sense how easy it will be to remove them later

It will be trivial to remove them later. The adapter will soon be the _only_ code even aware of their existence, so they should be easy to strip out entirely in a very small change set.

joxl commented:
> This seems to be the crux of the update process

You're absolutely right.

> Do you have a sense how this command will work?

I have the design worked out but not implemented. This will come after the multiplexing implementation because it depends on how that logic handles differences between the primary and secondary indexes (e.g. multiplexer gets a `DELETE` request, doc exists in primary but not secondary yet, etc).  This command will initiate the actual `/reindex` API request, monitor that task's progress and verify both indexes are equal before returning and reporting the success. Kind of like `rsync`, but for an Elastic index, and with fewer options, haha.

joxl commented:
In response to @calellowitz's three comments ([1][1], [2][2] and [3][3]):

[1]: https://github.com/dimagi/commcare-hq/pull/31125#discussion_r824145523
[2]: https://github.com/dimagi/commcare-hq/pull/31125#discussion_r824147121
[3]: https://github.com/dimagi/commcare-hq/pull/31125#discussion_r824150016

> i think `get` or the previous `get_doc` is a nicer name since it matches similar concepts in other parts of our code

> we use the phrase `iter_docs` for the same concept in other parts of our code

> the older name fits better into our naming conventions, where similar operations are typically called `get_docs`

The `iter_fetch()` method is a "synthetic" HQ function (it chunks up multiple "multi get" requests), and it makes much more sense to use HQ conventional `iter_docs` name as suggested.

My original (as in new) design was to rally around the concept of "fetch" which is the common, underlying _functionality_ of these three methods; with the added bonus of it being much more grepable than any combination of "get, "doc" and "source". In hindsight I agree that keeping them more conventional is more valuable than the alternative.

I'll push these changes soon.

joxl committed:
Code review: rename `fetch_many` to `get_docs`

joxl committed:
Code review: rename `iter_fetch` to `iter_docs`

joxl committed:
Code review: rename `fetch` to `get`

joxl commented on corehq/apps/es/client.py:
```diff
@@ -0,0 +1,877 @@
+                docs.append(doc_result["_source"])
+        return docs
+
+    def iter_fetch(self, doc_ids, chunk_size=100):
```
Renamed to `iter_docs` in 2e2881c2992

joxl commented on corehq/apps/es/client.py:
```diff
@@ -0,0 +1,877 @@
+            query.pop(extra, None)
+        return query
+
+    def fetch_many(self, doc_ids):
```
Renamed `get_docs` in 66741e28d8f

joxl commented on corehq/ex-submodules/pillowtop/utils.py:
```diff
@@ -237,30 +249,33 @@ def _is_deleted(change):
+            #
+            # The previous implementation didn't have an 'else' clause here,
+            # meaning if this did happen, then an action lacking an '_op_type'
+            # key was added to the payload . I assume Elastic would return an
```
Leaving the existing behavior will be tricky since the new BulkActionItem design is quite robust and doesn't support the caller "tweaking" the bulk payload in this way.  I'll do some testing and digging into these assumptions to determine exactly what happens in this scenario with the existing behavior. Maybe once I know that it will be more obvious how the existing behavior can be maintained.

joxl replied:
This is worse than I originally concluded.  If it is possible for an `else` condition to occur in the existing implementation, then it is resulting in a non-error Elasticsearch bulk action that is certainly not what we want.

Existing code [here](https://github.com/dimagi/commcare-hq/blob/a3ea4555770/corehq/ex-submodules/pillowtop/utils.py#L240-L263).

If the neither `if` or `elif` conditions are met, the result is an action being added to the payload that looks like this:

```python
{
    "_index": index_info.alias,
    "_type": index_info.type,
    "_id": change.id,
}
```

The [Elasticsearch bulk docs](https://elasticsearch-py.readthedocs.io/en/2.4.0/helpers.html#bulk-helpers) specify that an action lacking the `_op_type` key defaults to an `index` action.  This means that the above, sparse bulk action item will  always perform an `index` action for document `_id = change.id`. This will do one of:

1. index a new, empty (i.e. `_source = {}`) document which may or may not be using a valid `_id`, or
2. clear all properties on an existing document with that `_id`.

I tested both scenarios using the `bulk()` function and confirmed this is indeed what happens.

Changing the new BulkActionItem design to replicate this behavior is almost certainly not what we want to do. Thoughts?

joxl replied:
I submitted #31243 to fix this in order to keep this adapter PR free of Elastic processing logic changes.

joxl commented:
Thanks for your valuable contributions here, @calellowitz.

> Can you explain a bit more how this relates to the multiplexing work?

In order for multiplexing to work, all Elastic document-modifying calls must go through the adapter (i.e. no parts of HQ can be allowed to reach "around" the adapter and use the low-level third-party client object directly).  This is required in order for the adapter to be able to definitively know how to translate a "primary" Elastic action into a second action that is meaningful on the transient, secondary index (which may be in any state of missing, partially populated or fully synchronized). In other words, if there are parts of HQ that are changing Elastic documents outside the knowledge of the adapter, there is no way for it to know (with certainty) how to appropriately perform updates in the secondary index.

> Do you have a sense how far along this work is?

The next steps are to update Elastic-writing code to use the adapter directly rather than the (now) proxy-object `ElasticsearchInterface`.  That change set will be (again) far-reaching, and I plan to split it up into smaller PRs for easier review (this one got unwieldy, I do apologize for that).  The benefit of these upcoming changes are that it will decrease the complexity of working with Elasticsearch in HQ because it will:

1. Reduce client and index-related boilerplate that's duplicated throughout the codebase.
2. Allow any part of HQ that needs to use Elasticsearch to do so without also requiring at least some level of intimate knowledge on how pillowtop works (most relevant for future Elastic upgrades, but also important for other areas of code that use Elasticsearch without caring to know how or why pillowtop handles change processing).

> are the future plans for this work available anywhere I and others can see them?

Not at the moment. I'm waiting to continue that work until this adapter implementation completes review.  Documentation on the specifics of the multiplexing implementation and how it achieves data integrity are forthcoming and will be provided as early as possible.

> Do you have a plan to communicate these changes to the broader team?

Yes, in fact this is one of the more important aspects of this work: ensuring that the task of "working with Elasticsearch in HQ" is easily digestible by anyone who needs to do so.  Each PR that replaces `ElasticsearchInterface` with new adapter logic will include comprehensive documentation including code examples demonstrating "old vs new" ways of performing an Elasticsearch action.

calellowitz commented on corehq/apps/es/transient_util.py:
```diff
@@ -0,0 +1,103 @@
+    ElasticDocumentAdapter._report_and_fail_on_shard_failures(search_result)
+
+
+def _populate_doc_adapter_map(is_test):
```
nit: I just noticed that this function is only called outside this file. The `_` naming convention is used to indicate private functions that are not intended to be called outside the module. Since this is exclusively used outside the module, `populate_doc_adapter_map` would be a clearer name.

joxl replied:
Yes, I did that intentionally to discourage its use anywhere else. In hindsight this certainly has a smell. I'll remove the `_` and add a docstring instead.

calellowitz commented on corehq/ex-submodules/pillowtop/es_utils.py:
```diff
@@ -71,7 +71,10 @@ def _get_analysis(*names):
-# the value to None in settings.ES_SETTINGS
+# the value to this constant in settings.ES_SETTINGS
+# TODO: change this constant to an object that isn't actually a meaningful
+# Elastic index settings value (`None` is how you revert a setting to its
```
Thank you for the explanation. I understand the difference, but am unsure of the use case

>  The use case this could be important is if the commcare-cloud repo configures a cluster with specific "HQ default Elastic values", which may differ from the "Elastic default for that version". 

Settings values are set per environment in `commcare-cloud`, it would be an extremely confusing practice to explicitly set a value in `commcare-cloud` then unset in our hq settings (also configured in a different place in `commcare-cloud`), and I am not sure that is something we should be encouraging.

joxl replied:
Can you provide an example of the behavior you think is being encouraged? I'm not sure I follow.

In any case, there's no functional change here. I'll be sure to draw your attention to it if this changes in the future so you have an actual implementation to analyze.

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

joxl committed:
Update `build_bulk_payload` tests for `BulkActionItem` implementation

joxl committed:
Code review: don't use private name for non-private function

joxl committed:
Code review: don't pass `settings.UNIT_TESTING` as an argument

millerdev commented on corehq/apps/es/client.py:
```diff
@@ -812,6 +812,11 @@ def is_index(self):
 
+    def __eq__(self, other):
+        if not isinstance(other, BulkActionItem):
+            return False
```
nit: I think this could/should return [`NotImplemented`](https://docs.python.org/3/library/constants.html#NotImplemented).

The docs somewhat confusingly say "If all attempts return `NotImplemented`, the interpreter will raise an appropriate exception." It will return `False` if both `__eq__` methods return that:
```py
>>> class A:
...     def __eq__(self, other):
...         return NotImplemented
... 
>>> class B:
...     def __eq__(self, other):
...         return NotImplemented
... 
>>> A() == B()
False
```

joxl replied:
Thanks, I've never dug into doing this elegantly, but have always been curious about the various "correct" things to do here.

joxl committed:
Code review: improve `BulkActionItem.__eq__()` method

snopoke reviewed (commented):
There are a number of things I do think are good in this PR:
* pulling all the ES code together
* standardising the interface

I also don't have any specific code level objections.

But I'm still finding it quite hard to see the bigger picture, particularly with regard to multiplexing and reindexing.

As it stands this PR is only a partial solution to the points above and while I can see how those will be achieved if those were the only goals I don't think the level of changes here would justify that. Given that a major goal of this work is allow multiplexing and no information has been provided as to how that will be accomplished I don't think I can approve the changes as they stand.

I would prefer putting a hold on this work until the larger design has been reviewed.

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
```
@joxl It would be great to understand more about how these goals will be accomplished.

joxl replied:
> 3rd parties will be forced to reindex in the timeframes that we dictate.

This is not new. 3rd parties are already forced to reindex on our schedule. All but one or two Elasticsearch index names (which dictates a reindex) live in pillowtop code. 3rd parties have no control over this.

> to create the new index and then again to remove the new index.

New index creation does not require a manual step -- the multiplexing adapter will create the index when it starts up (just like pillows does now). The steps are 1) idempotent "sync" command and 2) delete the old index.  The current reindex process has these same "functional" steps, only the first is currently a post-deploy operation. The second (delete the old index) is currently a manual step AFAIK, so that's not new.

> Will that also be required for 3rd party hosters?

Yes, the same changelog process we use now to guide 3rd party hostsers through reindex operations will be used for multiplexing reindex operations.  There will also be safeguards in place to prevent accidentally missing steps and putting the Elastic cluster in an inconsistent state. Current reindex procedures do not have these sort of automatic safety measures.

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
+
+The least disruptive way to accomplish a multiplexing design is with an adapter
```
Can you speak more to how the adapter layer will allow us to do multiplexing. It's not clear from the current code who that will be accomplished.

joxl replied:
The adapter is a conduit through which all Elastic document actions go through (`index`, `update`, `delete`, `search`, etc). When the adapter is configured in "multiplexing mode", it will be able to translate every document action it receives into two actions: the "original" action is executed verbatim on the primary index, and an equivalent action is performed on the secondary index.  Because the adapter knows that _all_ actions for that index have come through it, it can thereby guarantee the integrity of documents that it maintains in the secondary index.

snopoke replied:
> When the adapter is configured in "multiplexing mode"

This is the crux of my question. How will this be done? I had a discussion with @millerdev about this last week and he indicated that this would be achieved by creating a delegating adapter that would delegate to multiple other adapters. Is this how you intend to make multiplexing work?

joxl replied:
Yes.

snopoke replied:
Would be great to add that into the doc.

snopoke reviewed (commented):
I re-read your documentation and there are a number of clues / hints in there as to how this will ultimately work but I still would like more detail on a number of points. I've made inline comments in the README file.

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+client library and high-level HQ components which need to read/write data in an
+Elasticsearch index. Conveniently, HQ already has the initial framework for this
+layer (the ``ElasticsearchInterface`` class), so the adapter layer is not a new
+concept. The reason the ``ElasticsearchInterface`` implementation cannot
+accommodate multiplexing is because it is not widely adopted for low-level
+Elasticsearch operations (e.g. index creation and configuration), and it does
+not sufficiently decouple low-level Elasticsearch document manipulation logic
+from the high-level HQ code that uses it.
```
Many areas of Elastic-related code in HQ simply step around the `ElasticsearchInterface` object by using the low-level `elasticsearch-py` client object directly. Multiplexing requires that the intermediate adapter absolutely _know_ everything that happens to documents in an index (so it can intelligently replicate that action in a secondary index).

snopoke replied:
In understand that, my question is why create a new construct instead of make sure the existing one is used everywhere.

joxl replied:
Because the existing `ElasticsearchInterface` implementation is the wrong level of abstraction for multiplexing (it's too low level for this functionality).

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+secondary) before fully committing to abandoning one of them. Creating a copy of
+an index is the unavoidable nature of a reindex operation, and multiplexing
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
```
> In the past we've accomplished this with temporary pillows to 'top up' the new index and then again to top up the old index after the switchover.

We have not used this process on all indexes. As far as I can tell we've only used it on one: the `case_search` index, which has a much simpler pillow configuration and usage than say, `users`, or `forms`.  This process simply will not work for all indexes, and is why a new process is needed.  This old process also has no mechanism for verifying data integrity of the index after the "top up" is complete.

snopoke replied:
> This process simply will not work for all indexes

I don't see why not. Why is the complexity of the pillow an issue? It's simple enough to create temporary pillows that only use specific processors. In fact we still have most of the original single processor pillow configurations in code.

> This old process also has no mechanism for verifying data integrity of the index after the "top up" is complete.

How do you intend to do that with the new process?

joxl replied:
> Why is the complexity of the pillow an issue?

It's error-prone and hard for an operator to do properly.

joxl replied:
> How do you intend to do that with the new process?

By spot-checking documents between the old (primary) and new (secondary) indexes.

snopoke replied:
To me this feels like a critical piece, we've tried this before in some ways and without a lot of success.

There is plenty of theory on population sampling so it would be great to figure out some of the details here so that we can be confident it's going to work.

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -0,0 +1,220 @@
+- At any point in the process, the rollback procedure is a simple code change
+  (i.e. revert PR, deploy).
+- The operation responsible for populating the secondary index is idempotent
+  *and* decoupled from the index configuration, allowing it to undergo change
+  iterations without aborting the entire process (thereby losing reindex
```
Ideally there are none, but the model supports it. If the "sync and verify" tool comes back saying the indexes are not equivalent, it can be changed (fixed) and run again without urgency or discarding of existing sync progress.

snopoke replied:
>  "sync and verify" tool 

What tool is this?

joxl replied:
I was referring to the management command outlined in step **2.** of the reindexing procedure in the README doc:

```markdown
2. Execute an idempotent management command that updates the secondary index
   from its primary counterpart.
```

millerdev committed:
Update import for Django 3.2

joxl committed:
Add design details to README.rst

joxl committed:
Improve multiplexing description in README.rst

Co-authored-by: Simon Kelly <skelly@dimagi.com>

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -80,6 +88,113 @@ This reindex procedure is inherently safe because:
+
+   1. Executes a Elastic ``reindex`` request with parameters to populate the
+      secondary index from the primary, configured to not overwrite existing
+      documents in the target (secondary) index.
```
> this will take data from the old index ... not ... from the primary data sources

Correct. This will be performed via Elasticsearch's Reindex API.

joxl committed:
Outline "primary/secondary swap" step as intended part of the procedure.

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

joxl committed:
Prevent regression from settings updates

- `case_search` and `forms` support custom index names via
  localsettings.py

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

calellowitz commented on corehq/apps/es/README.rst:
```diff
@@ -159,7 +159,20 @@ Reindex Procedure Details
+   run a "live test" on the new (secondary) index while keeping the old
+   (primary) index up-to-date.
+
+   - Reconfigure the adapter by swapping the "primary" and "secondary" index
```
Same question about having this be environment specific. Depending on the size of these indices, these processes will take way different amounts of times to finish and rely on third party hosters with unknown deploy schedules having run specific management commands already. Is there a way to make this safer for them, so that these potentially dangerous commands don't happen automatically in environments we don't control?

joxl replied:
If we need to add "rapid" index swapping (primary/secondary), this design doesn't prohibit it (it can be added later as needed). For the sake of this design I'm viewing it as an optimization, and the initial implementation is designed to handle reindexing via migrations alone.

> Is there a way to make this safer for them, so that these potentially dangerous commands ...

The safety is guaranteed by the migrations framework. Updating and running the management commands will not be dangerous because such operations will expect a specific "existing" state. That is, the migration (and "sync" management command) will fail if the Elasticsearch index state is not what it expects.

calellowitz reviewed (approved):
Thanks for adding all this info. I think I understand the intent a lot better now.

I left a few comments. One is a question to make sure I understand the abstraction correctly. I don't think any of the other ones are blocking on this PR, as they are mostly about the processes described for future work, and not the direction of the changes, but I do think they are worth figuring out sooner rather than later (and ideally before that code is written).

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -44,13 +44,21 @@ outages while keeping both up-to-date.
+backend. It is below "index-level" logic, and does not implement index-specific
+functionality needed in order for some indexes to be handled differently than
+others (e.g. some indexes are indexed individually while others are
+multiplexed). The document adapter implementation is a **document** abstraction
```
> Is it fair to say _document_ and _index_ are being used interchangeably here?

Hmm, sort of. In the sense that there is only one _Elasticsearch_ "document type" (i.e. mapping) in each index (sidenote: there is only one per index in the HQ implementation, but this will be enforced by Elasticsearch in newer versions). The word "document" here refers to an "Elasticsearch document", which the adapter creates with its `from_python()` method, so it's more than just an index.

> in that we use nested documents which are a level below these abstractions

I'm not sure what you're referring to here.

joxl commented on corehq/apps/es/README.rst:
```diff
@@ -80,6 +88,113 @@ This reindex procedure is inherently safe because:
+     the new (secondary) index. For example:
+
+     - creates the new index
+     - configures shards, replicas, etc for the index
```
> Much of that configuration does not exist in HQ

Yes, these would remain mostly the same, the only difference being that these changes<sup>1</sup> will now be _applied_ by migrations.  Just like now, there can be "default" settings defined in code, which pull in override values from settings.  The migration is the trigger that will apply those values. The implementation might<sup>2</sup> need a "smart" migration utility that can skip application of configurations that are defined in settings in the event that the cluster is already in the desired state.

<sup>1</sup> "these changes" refers to _only_ the settings that require a reindex.  Tunables that do not require a reindex can be managed outside of migrations.

<sup>2</sup> It's possible that the Elasticsearch API for these settings are idempotent on the server-side, in which case it would be safe for the migration to _always_ apply them.

snopoke reviewed (commented):
I think the points @calellowitz brought up about how these changes would affect 3rd party hosters are very relevant. In particular I think the process or rolling out ES index changes must be repeatable and flexible on any environment. For example, determining when it is safe for an environment to switch to the new index and allowing an env to have the same level of flexibility to switch back and forth between new and old indexes until they are ready to make it permanent. 

I think this is important not only for 3rd parties but for Dimagi as well since we have multiple environments that would need to apply these changes.

snopoke commented on corehq/apps/es/README.rst:
```diff
@@ -80,6 +88,113 @@ This reindex procedure is inherently safe because:
+      this step **2.** might be removed in favor of a blocking reindex during
+      the 2.4 --> 5.x upgrade.
+
+   3. Performs a cleanup operation on the secondary index to remove tombstone
```
Are these created by us or are they created by ES during the reindex process?

joxl replied:
Created by us. Tombstones are an artifact of the multiplexing design, not a native Elasticsearch feature.

calellowitz commented on corehq/apps/es/README.rst:
```diff
@@ -80,6 +88,113 @@ This reindex procedure is inherently safe because:
+     - configures shards, replicas, etc for the index
+     - sets the index mapping
+
+   - Review, merge and deploy this change.  At Django startup, the new
```
> The new reindexing workflow will be as safe and painless as any database migration. Reindexing might be more "complex" when it requires the extra "sync" management command, but there shouldn't be any reason to roll out a reindex slowly

For a more concrete example of my concern, what happens when we or, even more pressingly, a third party, needs to reindex because they want to add shards to an existing index? Each individual environment will need to reshard at various times, and it is likely that at some point that may be necessary for every commcare environment but not at the same time. If that operation requires code changes and added migrations, I fear that is impossible for third part hosters to do on their own (as it will both require them to fork the code base and generate migration conflicts for their reindex migrations), and difficult or painful for them (and maybe us) when we need to do that for a given environment.

joxl replied:
Third parties can't do that now, so I'm considering it "outside the scope of adapters". This design doesn't prohibit that feature, and I'm leaving it to be implemented at a later time.

snopoke replied:
> Third parties can't do that now

I don't think this is quite true. At least for the form and case search indexes the index names can be set via settings which means that an env could create a new index with a new name, use the existing reindex tools (`ptop_reindexer_v2` or `reindex_es_native`) to perform the reindex (or do it manually via the ES API) and then update the settings to point to the new index (and also switch the alias over).

We have certainly done that in the past on individual envs. I think the issue @calellowitz is addressing is that we are tying infrastructure concerns to the code and making it impossible to make transparent infra changes without also making code changes.

For changes that also require code changes e.g. changes to mappings, I think this approach is fine provided it is sufficiently flexible to allow individual envs to determine their own roll-out schedule (within some reasonable time frame).

snopoke replied:
To be clear, this 'hypothetical' scenario is exactly the scenario we were in with ICDS. The 'rec' project also has significant data and are in the process of migrating to their own infra. There are also discussions in USH about deploying CommCare to independent hardware for large EDSS projects (which hopefully won't happen but is not completely unlikely).

I don't think it should be very difficult to make it possible to override these things via settings. This would give us the flexibility to reindex envs independently if necessary.

It could be worth taking this to a doc to flesh out the exact reindex process step by step.

joxl commented:
@calellowitz, @snopoke I'm happy to continue discussing comments that are not directly relevant to the code changes on this PR, but I wanted to check in to see whether this was ready to merge for both of you given how it's cleared and has approvals.

snopoke commented:
> @calellowitz, @snopoke I'm happy to continue discussing comments that are not directly relevant to the code changes on this PR, but I wanted to check in to see whether this was ready to merge for both of you given how it's cleared and has approvals.

Fine with me.

joxl commented:
Offlined with @calellowitz for merge signoff.
